### PR TITLE
media-libs/opencv: Fix usage CPU instructions

### PR DESCRIPTION
**Pull request by ymuv**

### CODE REVIEW

**Comment by ymuv** [comment]
Bug report:
https://bugs.gentoo.org/show_bug.cgi?id=623614

**Comment by Amynka** [comment]
So if it does not fix it why don't you fix it? or what you mean?

**Comment by ymuv** [comment]
Because I am not good in ebuild system.
I don't know, how set  -DENABLE_OMIT_FRAME_POINTER=ON if **-fomit-frame-pointer** is in CFLAGS.

**Comment by Amynka** [comment]
Oki I will look into it :)

**Comment by Amynka** [comment]
One thing does it actually work how you changed it? I think it might be possibly all wrong what was in the ebuild and it cannot work at all.

**Comment by ymuv** [comment]
>One thing does it actually work how you changed it?
I look to ffmpeg ebuild ( https://github.com/gentoo/gentoo/blob/master/media-video/ffmpeg/ffmpeg-3.3.2.ebuild ).

>I think it might be possibly all wrong what was in the ebuild and it cannot work at all.
As for me, there was only problem with CPU flags.

**Comment by Amynka** [comment]
You didn't answer though what ffmpeg ebuild has is not really relevant to if it is actually working or not.

Because from what I can see all of -DENABLE_SSE= could not work at all.

**Comment by ymuv** [comment]
They didn't work, so I created PR to fix it.

**Comment by Amynka** [comment]
Yes exactly and I am asking if your solution work if you tested it......

**Comment by ymuv** [comment]
I test in x86 cpu with instructions: 
sse to sse4_2, popcnt.
CPU flags is OK after patch.

I can't test: arm_neon, avx, avx2, fma3, xop.

There is no need to fix **-DENABLE_FAST_MATH=OFF** because ffast-mat flag come from CFLAGS.

**Comment by ymuv** [comment]
I fix omit-frame-pointer.

**Comment by Amynka** [comment]
Coool thanks !

**Comment by ymuv** [comment]
@Amynka, can You merge it? 
This PR tested on x86 instruction (SSE-SSE4.2) (I can't test on arm CPU and X86 AVX).

**Comment by Amynka** [comment]
I am running tests on it when they end i will merge it yes :)

**Comment by ymuv** [comment]
>You are using function get-flag which is in flag-o-matic eclass but you are not inheriting the eclass. I need to take closer look at it.

I try:
1. compile with
`CFLAGS="-march=westmere -mtune=westmere -maes -O2 -pipe -mmmx -msse4.2 -ftracer -mfpmath=sse -msse -msse2 -mssse3 -msse4.1 -msse4.2 **-fomit-frame-pointer** -msse4.2 $CACHE $FLTO $FFAST_MATH $LOOP_FLAG"`

And check, that **-DENABLE_FAST_MATH=ON** and gcc compile with  **-fomit-frame-pointer**
Use simd counter script (**http://optimization.hardlinux.ru/wp-content/uploads/2013/03/simd_count.zip**) to check, that instructions are in .so libs. There are sse-sse4.1 instructions in shared libraries:
> ./simd_count /usr/lib/libopencv_imgproc.so
> nop:            10637
> call:           18199
> i686:           1830
> mmx:            7595
> sse:            22065
> sse2:           24093
> sse3:           29
> ssse3:          87
> sse4.1:         256
> 3dnowext:               589

2. Remove -fomit-frame-pointer, check that gcc compile without -fomit-frame-pointer (it use **-fomit-frame-pointer  ... -fno-omit-frame-pointer**)

3. Compile ebuild without patch. Run simd counter script  (**http://optimization.hardlinux.ru/wp-content/uploads/2013/03/simd_count.zip**) to check, that instructions are in .so libs. There is only sse instructions:

> ./simd_count /usr/lib/libopencv_imgproc.so
> nop:            9671
> call:           16998
> i686:           2598
> sse:            16091

File: media-libs/opencv/opencv-3.2.0-r1.ebuild [hunk 1]
```diff
@@ -233,20 +269,19 @@ src_configure() {
 		-DENABLE_SOLUTION_FOLDERS=OFF
 		-DENABLE_PROFILING=OFF
 		-DENABLE_COVERAGE=OFF
-		-DENABLE_OMIT_FRAME_POINTER=OFF
+		-DENABLE_OMIT_FRAME_POINTER=${ENABLE_OMIT_FRAME_POINTER}
```
Comment (Amynka) [inline media-libs/opencv/opencv-3.2.0-r1.ebuild]: You need to use usex here

Comment (ymuv) [inline media-libs/opencv/opencv-3.2.0-r1.ebuild]: -fomit-frame-pointer is not use flag, so I can't use it with usex.
From https://devmanual.gentoo.org/ebuild-writing/eapi/: 
>USAGE: usex <USE flag> [true output] [false output] [true suffix] [false suffix]]

I enter:

`$(usex $(get-flag -fomit-frame-pointer))`
And:
```
y-xeon ~ # emerge opencv -v
...
 * ERROR: media-libs/opencv-3.2.0-r1::gentoo failed (configure phase):
 *   USE Flag '-fomit-frame-pointer' not in IUSE for media-libs/opencv-3.2.0-r1
 * 
 * Call stack:
 *          ebuild.sh, line  115:  Called src_configure
 *        environment, line 6168:  Called usex '-fomit-frame-pointer'
 *   phase-helpers.sh, line  162:  Called use '-fomit-frame-pointer'
 *   phase-helpers.sh, line  200:  Called die
 * The specific snippet of code:
 *                              die "USE Flag '${u}' not in IUSE for ${CATEGORY}/${PF}"
```

File: media-libs/opencv/opencv-3.2.0-r1.ebuild [hunk 2]
```diff
@@ -131,6 +162,11 @@ src_configure() {
 	export ANT_OPTS+=" -Dfile.encoding=iso-8859-1"
 	java-ant-2_src_configure
 
+	ENABLE_OMIT_FRAME_POINTER=OFF
+	if [ ! -z $(get-flag -fomit-frame-pointer) ]; then
```
Comment (Amynka) [inline media-libs/opencv/opencv-3.2.0-r1.ebuild]: You have to inherit flag-o-matic eclass if you use the function from that eclass. Why not [[ ]] ? This whole piece of code is shaky..

Comment (ymuv) [inline media-libs/opencv/opencv-3.2.0-r1.ebuild]: I don't know that is  flag-o-matic, eclass.

File: media-libs/opencv/opencv-3.2.0-r1.ebuild (hunk 1 above)
Comment (Amynka) [inline media-libs/opencv/opencv-3.2.0-r1.ebuild]: Yes sorry my bad I read it badly

File: media-libs/opencv/opencv-3.2.0-r1.ebuild (hunk 2 above)
Comment (mgorny) [inline media-libs/opencv/opencv-3.2.0-r1.ebuild]: The whole concept looks wrong. You should patch the build system not to add any omitfp flags instead.

File: media-libs/opencv/opencv-3.2.0-r1.ebuild [hunk 3]
```diff
@@ -233,20 +269,19 @@ src_configure() {
...
-		-DENABLE_SSSE3=OFF
-		-DENABLE_SSE41=OFF
-		-DENABLE_SSE42=OFF
-		-DENABLE_POPCNT=OFF
-		-DENABLE_AVX=OFF
-		-DENABLE_AVX=OFF
-		-DENABLE_AVX2=OFF
-		-DENABLE_FMA3=OFF
-		-DENABLE_NEON=OFF
+		-DENABLE_SSE=$(usex cpu_flags_x86_sse)
```
Comment (mgorny) [inline media-libs/opencv/opencv-3.2.0-r1.ebuild]: Using those flags to control -m options is wrong. They are only for assembly and similar cases.

**Comment by Amynka** [comment]
I wonder does opencv 3.1.0-r7 works for you as you need? Maybe enough will be just to remove all those flags.

File: media-libs/opencv/opencv-3.2.0-r1.ebuild (hunk 2 above)
Comment (mudler) [inline media-libs/opencv/opencv-3.2.0-r1.ebuild]: also, you probably want to look at [usex](https://devmanual.gentoo.org/ebuild-writing/eapi/) from devmanual

File: media-libs/opencv/opencv-3.2.0-r1.ebuild (hunk 3 above)
Comment (ymuv) [inline media-libs/opencv/opencv-3.2.0-r1.ebuild]: Fix.

**Comment by Amynka** [comment]
I totally like the new way just with the patch.

**Comment by ymuv** [comment]
@mgorny , can You check PR?

**Comment by mgorny** [comment]
Thanks, it's certainly better. However, you seem to have missed SSE2 and older flags. Could you tackle them as well?

Also, I'm not sure if it wouldn't make the patch simpler if you didn't remove all the code rather than 'else' branches only. In that case, you could also remove the respective `-DENABLE_*=OFF` from the ebuild.

**Comment by Amynka** [comment]
I agree with the changes and I tested it and it works.. as mgorny said dont we wanna disable also sse2?

**Comment by Amynka** [comment]
Done. Thank you

**Pull request closed without merging**